File: python/sum-of-root-to-leaf-binary-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sumRootToLeaf(self, root: TreeNode) -> int:
        def getSum(self, node, num):
            if node==None: return
            if node.left==None and node.right==None:
                self.res+=num*2+node.val
            getSum(self, node.left, num*2+node.val)
            getSum(self, node.right, num*2+node.val)
        self.res=0
        getSum(self, root, 0)
        return self.res

File: cpp/sum-of-root-to-leaf-binary-numbers.cpp
1
Easy

21

36

Favorite

Share
Given a binary tree, each node has value 0 or 1.  Each root-to-leaf path represents a binary number starting with the most significant bit.  For example, if the path is 0 -> 1 -> 1 -> 0 -> 1, then this could represent 01101 in binary, which is 13.

For all leaves in the tree, consider the numbers represented by the path from the root to that leaf.

Return the sum of these numbers.

 

Example 1:



Input: [1,0,1,0,1,0,1]
Output: 22
Explanation: (100) + (101) + (110) + (111) = 4 + 5 + 6 + 7 = 22
 

Note:

The number of nodes in the tree is between 1 and 1000.
node.val is 0 or 1.
The answer will not exceed 2^31 - 1.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumRootToLeaf(TreeNode* root) {
        int res=0;
        dfs(root, res, 0);
        return res;
    }
private:
    void dfs(TreeNode* root, int& res, int num) {
        if (root==NULL) return;
        if (root->left==NULL && root->right==NULL) {
            res+=num*2+root->val;
            return;
        }
        dfs(root->left, res, num*2+root->val);
        dfs(root->right, res, num*2+root->val);
    }
};
